library-best-practices/JLBP-6.md

```diff
@@ -0,0 +1,141 @@
+- *If* the Java package is changed, use different maven coordinates (for example
+  different artifact id).
+- *If* the Java package is kept the same, use the same maven artifact ID.
+- In other words, rename the Java package if and only if the maven coordinates
+  change.
+
+Background
+----------
+
+When a library B in Java depends on another library A using Apache Maven, there
+are two handles that library B needs to use to access library A:
+
+1. The maven coordinates, following the form
+   `${group}:${artifactId}:${version}`, for example
+   `com.google.guava:guava:26.0-jre`. The maven coordinates are used to pull the
+   library's jar from Maven Central. Given a group and artifact id, Maven will
+   choose only one version to pull in.
+2. The Java package (the package used in `import` statements), for example
+   `com.google.common.collect.ImmutableList`. The Java package is used to select
+   a particular class within the jar pulled from Maven Central.
+
+Now when breaking changes are introduced to library A between major version 1
+and major version 2, a choice needs to be made: to rename or not rename? This
+question applies to both items listed above, the maven coordinates (1) and the
+Java package (2). Consider the following renaming scenario:
+
+- The maven coordinates start as `g1:a1:1.0.0`
+  - With no rename, the new major version's maven coordinates are
+    `g1:a1:2.0.0`. (Only one of the versions can be present in the dependency
+    tree.)
+  - With a rename, the new major version's maven coordinates are
+    `g2:a2:2.0.0`. (Both versions can be present in the dependency tree.)
+- The Java package starts as `com.google.a1`
+  - With no rename, the new major version's Java package is still
+    `com.google.a1`
+  - With a rename, the new major version's Java package is `com.google.a2`
+
+Given this scenario, here are the possible combinations of renamings:
+
+- **Don't rename Java package**:
+  - **Don't rename Maven coordinates (case 1)**: This approach can result in
+    diamond dependency conflicts because different branches of a dependency tree
+    can depend on different major versions, and only one of them can be pulled
+    in by maven in a particular dependency tree. Users are forced to change
+    their code between versions.
+  - **Rename maven coordinates (case 2)**: Users can easily pull in both the old
+    jar (`g1:a1:1.0.0`) and the new jar (`g2:a2:2.0.0`) accidentally through
+    transitive dependencies, because maven treats the two artifacts as distinct,
+    but since they have classes with the same Java packages, they are loaded in
+    unpredictable ways, leading to runtime exceptions (for example
+    `ClassNotFoundException`). It is difficult for users to ensure their build
+    tree only includes one of the artifacts. **NEVER DO THIS**. Example: guava
+    vs guava-jk5.
+- **Rename Java package**:
+  - **Don't rename Maven coordinates (case 3)**: The classes from `g1:a1:1.0.0`
```

Comment:
does junit do this? I think that latest junit jar includes the old 3.x packages too. Worth checking.

Reply:
If the old version is bundled with the new version, this has the same impact as described in case 4 (they can be used side by side). I will figure out how to integrate this idea.